assets/src/edit-story/components/library/libraryUploadDropTarget.js

```diff
@@ -34,16 +34,41 @@ import {
   UploadDropTargetMessageOverlay,
 } from '../uploadDropTarget';
 import { useUploader } from '../../app/uploader';
+import { useMedia } from '../../app/media';
+import {
+  getResourceFromLocalFile,
+  getResourceFromUploadAPI,
+} from '../../components/library/panes/media/mediaUtils';
 
 const MESSAGE_ID = 'edit-story-library-upload-message';
 
 function LibraryUploadDropTarget({ children }) {
   const { uploadFile } = useUploader();
+  const {
+    state: { media, mediaType, searchTerm },
+    actions: { fetchMediaSuccess },
+  } = useMedia();
   const onDropHandler = useCallback(
-    (files) => {
-      files.forEach(uploadFile);
+    async (files) => {
+      const localFiles = files.map(getResourceFromLocalFile);
+      const localMedia = await Promise.all(localFiles);
+      const filesUploading = files.map((file) => uploadFile(file));
+      fetchMediaSuccess({
+        media: [...localMedia, ...media],
+        mediaType,
+        searchTerm,
+      });
+      Promise.all(filesUploading).then((uploadedFiles) => {
+        const uploadedMedia = uploadedFiles.map(getResourceFromUploadAPI);
+
+        fetchMediaSuccess({
+          media: [...uploadedMedia, ...media],
```

Comment:
This is clever. But we also need to update the dependent elements, in case someone started using them right away. Right? I think this is a similar issue that @spacedmonkey is also trying to address in the #553.

Reply:
Interesting. What do you mean about dependent elements? Will not this Action update the Store itself and propagate the updates for those dependent elements?

Reply:
This is now #615. But basically, if we add an element in the story and then we update the underlying resource, we may want to run through the story and find all dependent elements and update them.